File: leetcode/python/hard/1335_minimum_difficulty_of_job_schedule.py
from functools import lru_cache


class Solution:
    def minDifficulty(self, jobDifficulty: list[int], d: int) -> int:
        n = len(jobDifficulty)
        if n < d:
            return -1

        base = [jobDifficulty[-1]] * len(jobDifficulty)
        for i in range(n-2, -1, -1):
            base[i] = max(base[i+1], jobDifficulty[i])
        @lru_cache(None)
        def dp(i: int, d: int) -> int:
            if d == 1:
                return base[i]
            res = float('inf')
            curMax = jobDifficulty[i]
            for j in range(i, n-d+1):
                curMax = max(curMax, jobDifficulty[j])
                res = min(res, curMax + dp(j+1, d-1))
            return res
        return dp(0, d)

s = Solution()
print(s.minDifficulty(jobDifficulty = [6,5,4,3,2,1], d = 2))
print(s.minDifficulty(jobDifficulty = [9,9,9], d = 4))
print(s.minDifficulty(jobDifficulty = [1,1,1], d = 3))

File: leetcode/rust/src/1335_minimum_difficulty_of_job_schedule.rs
use std::collections::HashMap;

impl Solution {
    pub fn min_difficulty(job_difficulty: Vec<i32>, d: i32) -> i32 {
        if d as usize > job_difficulty.len() {
            return -1;
        }

        let mut task = JobSchedule::new(job_difficulty, d as usize);
        task.min_difficulty(0, d as usize)
    }
}

struct JobSchedule {
    cache: HashMap<(usize, usize), i32>,
    base: Vec<i32>,
    job_difficulty: Vec<i32>,
    days: usize,
}

impl JobSchedule {
    pub fn new(job_difficulty: Vec<i32>, days: usize) -> Self {
        let mut base = vec![job_difficulty.last().unwrap().clone(); job_difficulty.len() + 1];
        for i in (0..job_difficulty.len()-1).rev() {
            base[i] = base[i + 1].max(job_difficulty.get(i).unwrap().clone());
        }
 
        Self {
            cache: HashMap::new(),
            base,
            job_difficulty,
            days,
        }
    }

    pub fn min_difficulty(&mut self, i: usize, days: usize) -> i32 {
        if days == 1 {
            return self.base[i];
        }

        if let Some(&v) = self.cache.get(&(i, days)) {
            return v;
        }

        let mut res = std::i32::MAX;
        let mut max = 0;
        for j in i..self.job_difficulty.len() - days + 1 {
            max = max.max(self.job_difficulty[j]);
            res = res.min(max + self.min_difficulty(j + 1, days - 1));
        }

        self.cache.insert((i, days), res);
        res
    }
    
}